rmf_gazebo_plugins/src/readonly.cpp

```diff
@@ -0,0 +1,466 @@
+  _graph = _level.nav_graphs[_nav_graph_index];
+  RCLCPP_INFO(logger(), "Nav graph contains [%d] lanes" , _graph.edges.size());
+  for (const auto& edge : _graph.edges)
+  {
+    // Inserting entry for v1_idx
+    auto entry = _neighbor_map.find(edge.v1_idx);
+    if (entry == _neighbor_map.end())
+    {
+      // We assume the vertex is a neighbor of itself to account for when the
+      // robot turns around and heads back to the start index
+      std::unordered_set<std::size_t> neighbors({edge.v1_idx, edge.v2_idx});
+      _neighbor_map.insert(std::make_pair(edge.v1_idx, neighbors));
+    }
+    // Updating existing entry
+    else
+    {
+      entry->second.insert(edge.v2_idx);
+    }
+
+    // If bidrectional, add entry for v2_idx
+    if (edge.edge_type == edge.EDGE_TYPE_BIDIRECTIONAL)
+    {
+      entry = _neighbor_map.find(edge.v2_idx);
+      if (entry == _neighbor_map.end())
+      {
+        std::unordered_set<std::size_t> neighbors({edge.v1_idx, edge.v2_idx});
+        _neighbor_map.insert(std::make_pair(edge.v2_idx, neighbors));
+      }
+      // Updating existing entry
+      else
+      {
+        entry->second.insert(edge.v1_idx);
+      }
+    }
+
+  }
+
+  // Initialize _wp_names
+  for (std::size_t i =0; i < _graph.vertices.size(); ++i)
+    _wp_names.insert(std::make_pair(i, _graph.vertices[i].name));
+
+  _initialized_graph = true;
+
+}
+
+double ReadonlyPlugin::compute_ds(
+    const ignition::math::Pose3d& pose,
+    const std::size_t& wp)
+{
+  // TODO consider returning a nullptr instead
+  assert(_found_graph);
+  assert(wp < _graph.vertices.size());
+  
+  ignition::math::Vector3d world_position{pose.Pos().X(), pose.Pos().Y(), 0};
+  ignition::math::Vector3d graph_position{
+    _graph.vertices[wp].x,
+    _graph.vertices[wp].y,
+    0};
+
+  return std::abs((world_position - graph_position).Length());
+}
+
+void ReadonlyPlugin::initialize_start(const ignition::math::Pose3d& pose)
+{
+  if (_initialized_start)
+    return;
+
+  if (!_initialized_graph)
+    return;
+
+  bool found = false;
+  for (std::size_t i = 0; i < _graph.vertices.size(); i++)
+  {
+    if (_graph.vertices[i].name.c_str() == _start_wp_name)
+    {
+      found = true;
+      _start_wp = i;
+      RCLCPP_INFO(logger(), "Start waypoint found in nav graph");
+    }
+  }
+
+  // TODO find the closest wp if the coordiantes do not match
+  if (compute_ds(pose, _start_wp) < 1e-1)
+  {
+    _initialized_start = true;
+    // Here we initialzie the next waypoint
+    compute_path(pose);
+    RCLCPP_INFO(logger(), "Start waypoint successfully initialized");
+  }
+
+  else
+  {
+    RCLCPP_ERROR(
+      logger(),
+      "Spawn coordinates [%f,%f,%f] differs from that of waypoint [%s] in nav_graph [%f, %f, %f]",
+       pose.Pos().X(), pose.Pos().Y(),0,
+      _start_wp_name.c_str(),
+      _graph.vertices[_start_wp].x, _graph.vertices[_start_wp].y, 0);
+  }
+  
+}
+std::size_t ReadonlyPlugin::get_next_waypoint(const std::size_t& start_wp,
```

Comment:
passing a primitive type like `std::size` by value is more efficient than `const &`, optimizers can flex their muscles when handling primitive types by value

Reply:
Thanks for letting me know! Will make the change.